packages/core/lib/utils.js

```diff
@@ -5,6 +5,47 @@ const flatten = require('flat');
 const isPlainObject = require('is-plain-object');
 const escapeRegExp = require('escape-string-regexp');
 
+const isProduction = process.env.NODE_ENV === 'production';
+
+exports.invariant = (condition, message) => {
+  if (condition) {
+    return;
+  }
+  let error;
+  if (isProduction) {
+    error = new Error('Invariant failed');
+  } else {
+    error = new Error(`Invariant failed: ${message}`);
+  }
+  error.framesToPop = 1;
+  throw error;
+};
+
+exports.validate = (strategy, ...validators) => {
+  const validate = (data) => validators.forEach((fn) => fn(data));
+  return Object.defineProperty(
+    (functions, ...initialArgs) =>
+      strategy(
+        functions.map((fn) => (...callArgs) => {
+          validate({ before: true, callArgs, initialArgs });
+          const result = fn(...callArgs);
```

Comment:
I'd consistently go down the Promise path here for the sake of readability.

```diff
- const result = fn(...callArgs);
- if (result && typeof result.then === 'function') {
-   return result.then(function(result) {
-     validate({ after: true, result, callArgs, initialArgs });
-     return result;
-   });
- } else {
-   validate({ after: true, result, callArgs, initialArgs });
-  return result;
- }
+ Promise.resolve(fn(...callArgs)).then((result) => {
+   validate({ after: true, result, callArgs, initialArgs });
+   return result;
+ });
```

Reply:
I would certainly like that, but we cannot: we will want to add validation to `core` mixin methods and those are all synchronous.